### PR TITLE
[IDP-1412] Attempt to retrieve and cache JWTs in background

### PR DESCRIPTION
**Pull request by asimmon**
## Description of changes

- On application startup, we now try to retrieve client credentials JWTs for each registered authenticated HttpClient. This will put them in the cache for later use.
- When the library retrieves a brand new token, it creates a background task that will attempt to retrieve a new token a little bit before the existing token is removed from the cache.
- As a result, it's very likely that HTTP requests that use the client credentials grant will use a cached token instead of blocking to retrieve a new one.
- This change requires the running application to be using the generic host to have the JWTs loaded on startup.
- Added an option to disable this behavior.

## Breaking changes

Added `ClientCredentialsOptions.EnablePeriodicTokenBackgroundRefresh` to allow opting out from this new behavior.

If the running app isn't using the generic .NET host, we won't attempt to cache tokens on startup. However, we will still start background refresh tasks later when we successfully get a brand new token.

## Additional checks

The integration test has been modified to ensure that:
- Tokens are cached before that we even tried to make any HTTP request
- When a token expires, the background refresh process takes care of retrieving a new one even if we done make any authenticated HTTP request

Manual QA with a real client/server and identity provider
- Token is cached on startup
- Background refresh loop works - once broken because identity provider is down - it restarts when attempting a new authenticated request

### CODE REVIEW

File: src/Workleap.Extensions.Http.Authentication.ClientCredentialsGrant/CacheTokenOnStartupBackgroundService.cs [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+    protected override Task ExecuteAsync(CancellationToken stoppingToken)
+    {
+        this._clientNames.AddRange(this._backgroundServiceOptions.Value.ClientCredentialPoweredClientNames);
+
+        this._applicationStartedRegistration = this._applicationLifetime.ApplicationStarted.Register(() =>
+        {
+            foreach (var clientName in this._clientNames)
+            {
+                // We don't await this task because we want to cache all tokens in parallel (fire and forget)
+                _ = this.CacheTokenAsync(clientName, stoppingToken);
```
Comment (meziantou) [inline src/Workleap.Extensions.Http.Authentication.ClientCredentialsGrant/CacheTokenOnStartupBackgroundService.cs:38]: You should do an actual fire and forget. If there is an error, it would report an TaskUnhandledException. 

https://www.meziantou.net/fire-and-forget-a-task-in-dotnet.htm

File: src/Workleap.Extensions.Http.Authentication.ClientCredentialsGrant/ClientCredentialsTokenManagementService.cs [hunk 2]
```diff
@@ -17,16 +18,28 @@ internal class ClientCredentialsTokenManagementService : IClientCredentialsToken
     private readonly ConcurrentDictionary<string, Lazy<Task<ClientCredentialsToken>>> _lazyGetNewTokenTasks;
     private readonly IClientCredentialsTokenEndpointService _tokenEndpointService;
     private readonly IClientCredentialsTokenCache _tokenCache;
+    private readonly CancellationToken _backgroundRefreshCancellationToken;
 
-    public ClientCredentialsTokenManagementService(IClientCredentialsTokenEndpointService tokenEndpointService, IClientCredentialsTokenCache tokenCache)
+    private long _backgroundRefreshedTokenCount;
```
Comment (meziantou) [inline src/Workleap.Extensions.Http.Authentication.ClientCredentialsGrant/ClientCredentialsTokenManagementService.cs:23]: If this is only for testing purposes, I think an int or uint would be enough

File: src/Workleap.Extensions.Http.Authentication.ClientCredentialsGrant/ClientCredentialsTokenManagementService.cs [hunk 3]
```diff
@@ -59,4 +65,38 @@ private async Task<ClientCredentialsToken> SynchronizeAsync(string clientName, F
...
+
+        var cacheEvictionTime = await this._tokenCache.SetAsync(clientName, newToken, cancellationToken).ConfigureAwait(false);
+        var cacheDuration = cacheEvictionTime - DateTimeOffset.UtcNow;
+
+        if (cacheDuration > TimeSpan.Zero)
+        {
+            const double backgroundRefreshDelayFactor = 0.8;
+            var delayBeforeNextBackgroundRefresh = TimeSpan.FromTicks((long)Math.Round(cacheDuration.Ticks * backgroundRefreshDelayFactor));
+
+            _ = this.ScheduleTokenBackgroundRefreshAsync(clientName, delayBeforeNextBackgroundRefresh);
```
Comment (meziantou) [inline src/Workleap.Extensions.Http.Authentication.ClientCredentialsGrant/ClientCredentialsTokenManagementService.cs:81]: Same comment here about Fire and Forget. If `_backgroundRefreshCancellationToken` is cancelled while `Task.Delay`, you get an exception, so an TaskUnhandledException somewhere.

File: src/Workleap.Extensions.Http.Authentication.ClientCredentialsGrant/ClientCredentialsTokenManagementService.cs [hunk 4]
```diff
@@ -59,4 +65,38 @@ private async Task<ClientCredentialsToken> SynchronizeAsync(string clientName, F
             this._lazyGetNewTokenTasks.TryRemove(clientName, out _);
         }
     }
+
+    private Func<Task<ClientCredentialsToken>> GetNewTokenTaskFactory(string clientName, CancellationToken cancellationToken) => async () =>
+    {
+        var newToken = await this._tokenEndpointService.RequestTokenAsync(clientName, cancellationToken).ConfigureAwait(false);
```
Comment (meziantou) [inline src/Workleap.Extensions.Http.Authentication.ClientCredentialsGrant/ClientCredentialsTokenManagementService.cs:71]: If this call fails (exception), I think the background refresh loop is stopped as you will never call `ScheduleTokenBackgroundRefreshAsync` again. Is it what we want?

File: src/Workleap.Authentication.ClientCredentialsGrant.Tests/IntegrationTests.cs [hunk 5]
```diff
@@ -131,6 +158,30 @@ public async Task Real_Client_Server_Communication()
...
+
+            // At this point we're not making any new authenticated HTTP request,
+            // but there should be a background refresh task that should have retrieved a new token for us and it's not the same as the one we got after the background caching
+            Assert.Equal(1, tokenManagementService.BackgroundRefreshedTokenCount);
+            var tokenAfterFirstBackgroundRefresh = await tokenCache.GetAsync(invoiceReadClientName, cts.Token);
+            Assert.NotNull(tokenAfterFirstBackgroundRefresh);
+            Assert.NotEqual(tokenAfterStartupBackgroundCaching, tokenAfterFirstBackgroundRefresh);
+
+            // If we wait a little bit longer, the token should be refreshed again
+            await Task.Delay(tokenAfterFirstBackgroundRefresh.Expiration - DateTimeOffset.UtcNow, cts.Token);
```
Comment (meziantou) [inline src/Workleap.Authentication.ClientCredentialsGrant.Tests/IntegrationTests.cs:180]: I don't know how reliable this is (more generally, `Task.Delay` in a test). I wonder if you should wait for an event or a SemaphoreSlim as you did with `WaitForTokenCachingToCompleteAsync`

File: src/Workleap.Extensions.Http.Authentication.ClientCredentialsGrant/CacheTokenOnStartupBackgroundService.cs [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+            // Expected when the application is shutting down
+        }
+        catch
+        {
+            // We did our best to cache the token on startup. It will be retried when an HttpClient will attempt to make authenticated requests.
+        }
+    }
+
+    // This is meant to be used for integration tests only
+    internal async Task WaitForTokenCachingToCompleteAsync(CancellationToken cancellationToken)
```
Comment (meziantou) [inline src/Workleap.Extensions.Http.Authentication.ClientCredentialsGrant/CacheTokenOnStartupBackgroundService.cs:67]: I'm not sure of this implementation. In fact, this method can only be called once. If you call this method twice, only one will "succeed". I think you can replace the SemaphoreSlim with a [TaskCompletionSource](https://learn.microsoft.com/en-us/dotnet/api/system.threading.tasks.taskcompletionsource?view=net-8.0&WT.mc_id=DT-MVP-5003978)

File: src/Workleap.Authentication.ClientCredentialsGrant.Tests/IntegrationTests.cs (hunk 5 above)
Comment (asimmon) [reply src/Workleap.Authentication.ClientCredentialsGrant.Tests/IntegrationTests.cs:180]: I know timing isn't always reliable in tests... however here we really depend on how long tokens issued by IdentityServer are valid, and I didn't find a way to have a "hook" on their side to detect tokens are actually expired.

I tried to mitigate the issue by waiting way longer than needed (token is evicted from cache after X seconds but we wait X + eviction buffer Y which is 3 seconds + token is actually refreshed prior to this -2 seconds before)

File: src/Workleap.Extensions.Http.Authentication.ClientCredentialsGrant/CacheTokenOnStartupBackgroundService.cs [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+        this._backgroundServiceOptions = backgroundServiceOptions;
+        this._allTokensCachedSignal = new SemaphoreSlim(initialCount: 0, maxCount: 1);
+        this._clientNames = new List<string>();
+    }
+
+    protected override Task ExecuteAsync(CancellationToken stoppingToken)
+    {
+        this._clientNames.AddRange(this._backgroundServiceOptions.Value.ClientCredentialPoweredClientNames);
+
+        this._applicationStartedRegistration = this._applicationLifetime.ApplicationStarted.Register(() =>
```
[the remark this answers is not in the document]
Comment (asimmon) [reply src/Workleap.Extensions.Http.Authentication.ClientCredentialsGrant/CacheTokenOnStartupBackgroundService.cs:33]: 1. We know for sure ([host source code](https://github.com/dotnet/runtime/blob/v8.0.5/src/libraries/Microsoft.Extensions.Hosting/src/Internal/Host.cs#L159)) that when `IHostApplicationLifetime.ApplicationStarted` is triggered, the host is fully started (all IHostedService have been executed, web server is up and running)
2. If we were to execute this code directly in the `BackgroundService.ExecuteAsync` method, then the web server wouldn't be started yet. While this is not an issue with actual production code (Kestrel is not up yet, no problem), it does crash in integration tests because the testing server is not available yet but is requested by the integration tests which attempts to react to our "token are cached" and tries to access the web app (which is still starting at this point)

TLDR: it was required for integration test, but overall it's better to listen to this signal because we know for a fact that at this point the server is up and running

File: src/Workleap.Extensions.Http.Authentication.ClientCredentialsGrant/CacheTokenOnStartupBackgroundService.cs (hunk 6 above)
Comment (asimmon) [reply src/Workleap.Extensions.Http.Authentication.ClientCredentialsGrant/CacheTokenOnStartupBackgroundService.cs:67]: Oh yeah you're right, TCS are perfect for this use case. Made the change, thx.

File: src/Workleap.Extensions.Http.Authentication.ClientCredentialsGrant/ClientCredentialsTokenManagementService.cs (hunk 2 above)
Comment (asimmon) [reply src/Workleap.Extensions.Http.Authentication.ClientCredentialsGrant/ClientCredentialsTokenManagementService.cs:23]: Even though we only validate the value in tests, it's still going to be incremented in production environments. Very unlikely that we reach int.MaxValue or uint.MaxValue, but still.

File: src/Workleap.Extensions.Http.Authentication.ClientCredentialsGrant/ClientCredentialsTokenManagementService.cs (hunk 4 above)
Comment (asimmon) [reply src/Workleap.Extensions.Http.Authentication.ClientCredentialsGrant/ClientCredentialsTokenManagementService.cs:71]: Actually this is not true :)

If the background refresh loop is stopped because we failed to retrieve a token, then the next time an authenticated call is made and goes through our HttpClient delegated handlers, we will request a new token, then start the background task again => starting the loop.

I did QA that by shutting down my identity server. Token refresh was set to every 5 seconds. The background loop stopped. Started my identity server again, made an authenticated call, and background refresh loop started again.

File: src/Workleap.Extensions.Http.Authentication.ClientCredentialsGrant/ClientCredentialsTokenManagementService.cs [hunk 8]
```diff
@@ -59,4 +60,37 @@ private async Task<ClientCredentialsToken> SynchronizeAsync(string clientName, F
...
+
+        var cacheEvictionTime = await this._tokenCache.SetAsync(clientName, newToken, cancellationToken).ConfigureAwait(false);
+        var cacheDuration = cacheEvictionTime - DateTimeOffset.UtcNow;
+
+        if (cacheDuration > TimeSpan.Zero)
+        {
+            const double backgroundRefreshDelayFactor = 0.8;
+            var delayBeforeNextBackgroundRefresh = TimeSpan.FromTicks((long)Math.Round(cacheDuration.Ticks * backgroundRefreshDelayFactor));
```
[the remark this answers is not in the document]
Comment (asimmon) [reply src/Workleap.Extensions.Http.Authentication.ClientCredentialsGrant/ClientCredentialsTokenManagementService.cs:79]: Yeah I tried to use it, but unavailable in netstandard2.0

File: src/Workleap.Extensions.Http.Authentication.ClientCredentialsGrant/ClientCredentialsTokenManagementService.cs (hunk 3 above)
Comment (asimmon) [reply src/Workleap.Extensions.Http.Authentication.ClientCredentialsGrant/ClientCredentialsTokenManagementService.cs:81]: Done

File: src/Workleap.Extensions.Http.Authentication.ClientCredentialsGrant/CacheTokenOnStartupBackgroundService.cs [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+        {
+            var options = this._clientCredentialsOptionsMonitor.Get(clientName);
+            if (!options.EnablePeriodicTokenBackgroundRefresh)
+            {
+                return;
+            }
+
+            _ = await this._tokenManagementService.GetAccessTokenAsync(clientName, CachingBehavior.ForceRefresh, cancellationToken).ConfigureAwait(false);
+
+            if (Interlocked.Increment(ref this._successfulCachedTokenCount) == this._clientNames.Count)
```
Comment (Gcaya) [inline src/Workleap.Extensions.Http.Authentication.ClientCredentialsGrant/CacheTokenOnStartupBackgroundService.cs:60]: Given the `ClientsNames` list is populated directly from the `ClientCredentialPoweredClientNames` property of the 
`CacheTokenOnStartupBackgroundServiceOptions` options and in the registrations in `AddClientCredentialsHandler` directly add a registered httpClient to the list.

Isn't there a possibility that `this._allTokensCachedSignal.SetResult(true);` will never be called if one of the clients hasn't enabled `EnablePeriodicTokenBackgroundRefresh` cause then I feel the `_successfulCachedTokenCount` will never equal `_clientNames.Count` no?

File: src/Workleap.Extensions.Http.Authentication.ClientCredentialsGrant/CacheTokenOnStartupBackgroundServiceOptions.cs [hunk 10]
```diff
@@ -0,0 +1,6 @@
+﻿namespace Workleap.Extensions.Http.Authentication.ClientCredentialsGrant;
+
+internal sealed class CacheTokenOnStartupBackgroundServiceOptions
+{
+    public HashSet<string> ClientCredentialPoweredClientNames { get; } = new HashSet<string>(StringComparer.Ordinal);
```
Comment (Gcaya) [inline src/Workleap.Extensions.Http.Authentication.ClientCredentialsGrant/CacheTokenOnStartupBackgroundServiceOptions.cs:5]:
```suggestion
    public HashSet<string> ClientCredentialsPoweredClientNames { get; } = new HashSet<string>(StringComparer.Ordinal);
```

File: src/Workleap.Extensions.Http.Authentication.ClientCredentialsGrant/CacheTokenOnStartupBackgroundServiceOptions.cs [hunk 11]
```diff
@@ -0,0 +1,6 @@
+﻿namespace Workleap.Extensions.Http.Authentication.ClientCredentialsGrant;
+
+internal sealed class CacheTokenOnStartupBackgroundServiceOptions
```
Comment (Gcaya) [inline src/Workleap.Extensions.Http.Authentication.ClientCredentialsGrant/CacheTokenOnStartupBackgroundServiceOptions.cs:3]: How about?
```suggestion
internal sealed class OnStartupTokenCacheBackgroundServiceOptions
```

File: src/Workleap.Extensions.Http.Authentication.ClientCredentialsGrant/CacheTokenOnStartupBackgroundService.cs (hunk 9 above)
Comment (asimmon) [reply src/Workleap.Extensions.Http.Authentication.ClientCredentialsGrant/CacheTokenOnStartupBackgroundService.cs:60]: Yep you're right, I'll fix it. I added EnablePeriodicTokenBackgroundRefresh at the end and didn't consider that point.

File: src/Workleap.Extensions.Http.Authentication.ClientCredentialsGrant/CacheTokenOnStartupBackgroundServiceOptions.cs (hunk 11 above)
Comment (asimmon) [reply src/Workleap.Extensions.Http.Authentication.ClientCredentialsGrant/CacheTokenOnStartupBackgroundServiceOptions.cs:3]: I'll rename the service class too

File: src/Workleap.Extensions.Http.Authentication.ClientCredentialsGrant/CacheTokenOnStartupBackgroundServiceOptions.cs (hunk 10 above)
Comment (asimmon) [reply src/Workleap.Extensions.Http.Authentication.ClientCredentialsGrant/CacheTokenOnStartupBackgroundServiceOptions.cs:5]: Indeed!

File: src/Workleap.Extensions.Http.Authentication.ClientCredentialsGrant/CacheTokenOnStartupBackgroundService.cs [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+    protected override Task ExecuteAsync(CancellationToken stoppingToken)
+    {
+        this._clientNames.AddRange(this._backgroundServiceOptions.Value.ClientCredentialPoweredClientNames);
+
+        this._applicationStartedRegistration = this._applicationLifetime.ApplicationStarted.Register(() =>
+        {
+            foreach (var clientName in this._clientNames)
+            {
+                // We don't await this task because we want to cache all tokens in parallel
+                this.CacheTokenAsync(clientName, stoppingToken).Forget();
```
[the remark this answers is not in the document]
Comment (asimmon) [reply src/Workleap.Extensions.Http.Authentication.ClientCredentialsGrant/CacheTokenOnStartupBackgroundService.cs]: I can't do that because it doesn't have the same behavior - we can't consume the cancellation token passed in WaitForTokenCachingToCompleteAsync

**Review by Gcaya — APPROVED** [review]
